src/v/ssx/future-util.h

```diff
@@ -99,13 +99,88 @@ inline auto async_transform(Rng&& rng, Func&& func) {
     return async_transform(rng.begin(), rng.end(), std::forward<Func>(func));
 }
 
+/// \brief Run tasks synchronously in order and wait for completion only
+/// invoking futures one after a previous has completed (range version).
+/// Difference is that the results are expected to be a series of collections
+/// for which are then transformed into one single collection.
+///
+/// Given a range [\c begin, \c end) of objects, run \c func on each \c *i in
+/// the range, and return a \c future<> containing a \c std::vector<> of
+/// std::vector<> of values. These values are then unflattened into a single
+/// vector as the functions return value. In case any of the given tasks fails,
+/// one of the exceptions is returned by this function as a failed future.
+///
+/// \param begin an \c InputIterator designating the beginning of the range
+/// \param end an \c InputIterator designating the end of the range
+/// \param func Function to invoke with each element in the range (will be
+/// futurized if it doesn't return a \c future<> and expected to return a
+/// std::vector<> of values)
+/// \return a \c future<> containing a \c std::vector<>
+/// of the results of the function invocations that resolves when all the
+/// function invocations complete.  If one or more return an exception, the
+/// return value contains one of the exceptions.
+// clang-format off
+template<typename Iterator, typename Func>
+CONCEPT(requires requires(Func f, Iterator i) {
+    *i++;
+    { i != i } ->std::convertible_to<bool>;
+    seastar::futurize_invoke(f, *i).get0().begin();
+    seastar::futurize_invoke(f, *i).get0().end();
+})
+// clang-format on
+inline auto async_flat_transform(Iterator begin, Iterator end, Func&& func) {
```

Comment:
i think you should rename this something like `async_flat_for_each`, because it seems to me that an alternative implementation here might use `when_all_succeed` like

when_all_succeed(begin, end).then([](vector<vector<values>>) { return generic_flat_combine(...); }

the complexity below is because of iterative nature of do_for_each which we can expose through the name.

Reply:
I like it +1

Reply:
The reason I didn't do this is because I figured if we wanted to do that implementation we would name it, `parallel_flat_transform` to align with the `parallel` family of methods @BenPope has included in this file that work that way.

Reply:
coo